Improve PySide6 Support: swapBehavior for DoubleBuffer, BUTTONMAP as pyqt6, xfail test_context

psobolewskiPhD opened the pull request:
Fixes #2407 by using `SwapBehavior` to set DoubleBuffer
PYSIDE6_API is included in the existing PYQT6_API logic:
https://github.com/vispy/vispy/blob/8bf4f9e538dee5cb986779006b84f62eed81ed7c/vispy/app/backends/_qt.py#L281-L287

This should work for PySide6 going back to at least 6.2 (LTS)
I found the Qt 6.2 (LTS) docs. The swap behavior was already present.
https://doc.qt.io/qt-6.2/qsurfaceformat.html#SwapBehavior-enum
https://doc.qt.io/qtforpython-6.2/PySide6/QtGui/QSurfaceFormat.html#PySide6.QtGui.PySide6.QtGui.QSurfaceFormat.SwapBehavior

Edit: Also uses the same BUTTONMAP as PyQT6
Plus, xfails the shared context tests: test_context.py
Fixes https://github.com/vispy/vispy/issues/2410

djhoese commented:
So this drops support for older PySide versions (6.1 and 6.2 most likely), this is OK with me. With this single change to the `if` statement, do your tests from the related issue work now @psobolewskiPhD?

psobolewskiPhD commented:
> So this drops support for older PySide versions (6.1 and 6.2 most likely), this is OK with me. With this single change to the `if` statement, do your tests from the related issue work now @psobolewskiPhD?

pyside6 6.2 is still fine. the only question is 6.0 and 6.1. I can't install those or find any docs on them.
Alas, it turns out this solves the DoubleBuffer issue, but unmasks a new issue. When running turntable example and trying to interact with the object:
```
Traceback (most recent call last):
  File "/Users/piotrsobolewski/Dev/vispy/vispy/app/backends/_qt.py", line 515, in mouseReleaseEvent
    button=BUTTONMAP[ev.button()],
KeyError: <MouseButton.LeftButton: 1>
Traceback (most recent call last):
  File "/Users/piotrsobolewski/Dev/vispy/vispy/app/backends/_qt.py", line 515, in mouseReleaseEvent
    button=BUTTONMAP[ev.button()],
KeyError: <MouseButton.LeftButton: 1>
```

djhoese commented:
That's probably an update needed here:

https://github.com/vispy/vispy/blob/8bf4f9e538dee5cb986779006b84f62eed81ed7c/vispy/app/backends/_qt.py#L194

psobolewskiPhD commented:
> That's probably an update needed here:
> 
> https://github.com/vispy/vispy/blob/8bf4f9e538dee5cb986779006b84f62eed81ed7c/vispy/app/backends/_qt.py#L194

Thanks for the tip!
Fixed.

I went through everywhere there is `PYQT6_API` and checked the PySide6 docs. I think everything should be correct now. And I get no errors.

I also added pyside6 to the xfails for shared context, so now the app tests can run without set fault on pyside6 (note that run individually, `test_context.py` does pass on pyside6 if `se_shared_contexts()` is set to TRUE, but run as all app tests still segfaults.
This adresses https://github.com/vispy/vispy/issues/2410

djhoese commented on vispy/app/tests/test_context.py:
```diff
@@ -68,9 +70,9 @@ def check():
-            pytest.xfail("Context sharing is not supported in PyQt5 at this time.")
+        # pyqt5 does not currently support context sharing, pyside6 seg faults on app tests
+        if 'pyqt5' in c1.app.backend_name.lower() or 'pyqt6' in c1.app.backend_name.lower() or 'pyside6' in c1.app.backend_name.lower():
+            pytest.xfail("Context sharing is not supported in PyQt5, PyQt6, PySide6 at this time.")
```
Pyside2 also, right?

psobolewskiPhD replied:
I don't know and I can't test, because pyside2 seems borked on arm64.
Pyside6 the context sharing works and the test passes if run individually, but if you run the app suite then you get seg fault.

psobolewskiPhD replied:
Oh, of course you're right. I'll add the xfail for pyside2 then.

djhoese commented:
Thanks @psobolewskiPhD!
